gc/base/SparseAddressOrderedFixedSizeDataPool.cpp

```diff
@@ -0,0 +1,423 @@
+
+	/* Initialize _freeListPool J9Pool */
+	_freeListPool = pool_new(sizeof(MM_SparseHeapLinkedFreeHeader), 6, 0, 0, OMR_GET_CALLSITE(), OMRMEM_CATEGORY_MM, POOL_FOR_PORT(OMRPORTLIB));
+
+	_objectToSparseDataTable = hashTableNew(
+		OMRPORTLIB, OMR_GET_CALLSITE(),
+		11,
+		sizeof(J9ObjSparseDataTableEntry),
+		sizeof(uintptr_t),
+		0,
+		OMRMEM_CATEGORY_MM,
+		entryHash, 
+		entryEquals,
+		NULL,
+		NULL);
+
+	/* Initialize _heapFreeList with initial values */
+	_heapFreeList = createNewSparseHeapFreeListNode(sparseHeapBase, _approxLargestFreeEntry);
+
+	if ((NULL == _freeListPool) || (NULL == _objectToSparseDataTable) || (NULL == _heapFreeList)) {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_initialization_failure(sparseHeapBase, _freeListPool, _objectToSparseDataTable, _heapFreeList);
+		return false;
+	}
+
+	Trc_MM_SparseAddressOrderedFixedSizeDataPool_initialization_success(sparseHeapBase, _freeListPool, _objectToSparseDataTable, _heapFreeList);
+	return true;
+}
+
+void
+MM_SparseAddressOrderedFixedSizeDataPool::tearDown(MM_EnvironmentBase *env) 
+{
+	if (NULL != _freeListPool) {
+		pool_kill(_freeListPool);
+		_freeListPool = NULL;
+	}
+
+	if (NULL != _heapFreeList) {
+		freeAllSparseHeapFreeListNodes();
+		_heapFreeList = NULL;
+	}
+
+	if (NULL != _objectToSparseDataTable) {
+		hashTableFree(_objectToSparseDataTable);
+		_objectToSparseDataTable = NULL;
+	}
+}
+
+/**
+ * Teardown and free the receiver by invoking the virtual #tearDown() function
+ */
+void
+MM_SparseAddressOrderedFixedSizeDataPool::kill(MM_EnvironmentBase *env)
+{
+	tearDown(env);
+	env->getForge()->free(this);
+}
+
+#if defined(OMR_GC_DOUBLE_MAPPING_FOR_OSX_SPARSE_HEAP_ALLOCATION)
+struct J9PortVmemIdentifier*
+MM_SparseAddressOrderedFixedSizeDataPool::getIdentifierFromDataPtr(void *dataPtr)
+{
+	J9ObjSparseDataTableEntry lookupEntry;
+	J9ObjSparseDataTableEntry *entry;
+
+	lookupEntry.dataPtr = dataPtr;
+	entry = (J9ObjSparseDataTableEntry *)hashTableFind(_objectToSparseDataTable, &lookupEntry);
+	if (NULL == entry || (entry->dataPtr != dataPtr)) {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_findEntry_failure(dataPtr);
+		return NULL;
+	}
+
+	return entry->identifier;
+}
+
+void
+MM_SparseAddressOrderedFixedSizeDataPool::recordDoubleMapIdentifierForData(void *dataPtr, struct J9PortVmemIdentifier *identifier)
+{
+	J9ObjSparseDataTableEntry lookupEntry;
+	J9ObjSparseDataTableEntry *entry;
+
+	lookupEntry.dataPtr = dataPtr;
+	entry = (J9ObjSparseDataTableEntry *)hashTableFind(_objectToSparseDataTable, &lookupEntry);
+
+	if (NULL == entry || (entry->dataPtr != dataPtr)) {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_findEntry_failure(dataPtr);
+	} else {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_findEntry_success(dataPtr);
+		entry->identifier = identifier;
+	}
+}
+#endif /* OMR_GC_DOUBLE_MAPPING_FOR_OSX_SPARSE_HEAP_ALLOCATION */
+
+bool
+MM_SparseAddressOrderedFixedSizeDataPool::addObjectDataEntrytoTable(void *dataPtr, void* proxyObjPtr, uintptr_t size)
```

Comment:
Think about some names that better describe what this accomplishes from a higher point of view (rather than talking about mechanics that occur within the method):
For example: mapSparseDataPtrToHeapObject. Then we would also have: findHeapObjectForSparseDataPtr etc.